File: content/templates/python-starter/helpers/shapes.py
"""A very small shapes library: one base class and two shapes."""

import math


class Shape:
    """Base class. Every shape has a name and can report its area."""

    def __init__(self, name):
        self.name = name

    def area(self):
        # Each subclass gives its own answer.
        return 0.0

    def describe(self):
        return f"{self.name}: area = {self.area():.2f}"


class Circle(Shape):
    def __init__(self, radius):
        super().__init__("Circle")  # let Shape store the name
        self.radius = radius

    def area(self):
        return math.pi * self.radius**2


class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__("Rectangle")
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

File: content/templates/csharp-starter/Shapes.cs
using System;

/// <summary>A very small shapes library: one base class and two shapes.</summary>
class Shape
{
    // A read-only property: other classes can read Name, only this one sets it.
    public string Name { get; }

    // A constructor builds the object.
    public Shape(string name)
    {
        Name = name;
    }

    // virtual means a subclass may replace this method with its own version.
    public virtual double Area()
    {
        return 0.0;
    }

    public string Describe()
    {
        return $"{Name}: area = {Area():F2}";
    }
}

/// <summary>A Circle is a Shape. ": Shape" reuses everything Shape has.</summary>
class Circle : Shape
{
    private readonly double radius;

    public Circle(double radius) : base("Circle") // let Shape store the name
    {
        this.radius = radius;
    }

    // override takes over Area() for Circles.
    public override double Area()
    {
        return Math.PI * radius * radius;
    }
}

class Rectangle : Shape
{
    private readonly double width;
    private readonly double height;

    public Rectangle(double width, double height) : base("Rectangle")
    {
        this.width = width;
        this.height = height;
    }

    public override double Area()
    {
        return width * height;
    }
}
